OpenSim/Common/Delay.h

```diff
@@ -0,0 +1,179 @@
+ *
+ * For times prior to the start of a simulation this Component behaves as
+ * though the input value had been constant at its initial value.
+ *
+ * @tparam T The type of the quantity to be delayed. Common choices are a
+ * SimTK::Real (see ScalarDelay) or a SimTK::Vector (see VectorDelay).
+ *
+ * @see ScalarDelay, VectorDelay
+ *
+ *
+ * This Component has a single input named "input" and a single output named
+ * "output". The typical way to use this Component is as a subcomponent in
+ * another component. Here is a basic example in which a controller specifies
+ * the control signal for a single actuator using the delayed value of a
+ * coordinate:
+ * @code
+ * using namespace OpenSim;
+ * using namespace SimTK;
+ * class MyController : Controller {
+ * OpenSim_DECLARE_CONCRETE_OBJECT(MyController, Controller);
+ * public:
+ *     OpenSim_DECLARE_PROPERTY(delay, double, "Duration of delay (seconds).");
+ *     MyController() { constructProperties(); }
+ *     void computeControls(const State& s, Vector &controls) const override {
+ *         double q = _coordDelay.getValue(s);
+ *         // Delayed negative feedback control law.
+ *         double u = -10 * q;
+ *         getModel().getActuatorSet()[0].addInControls(Vector(1, u), controls);
+ *     }
+ * private:
+ *     void constructProperties() {
+ *          constructProperty_delay(0.01); // 10 milliseconds
+ *     }
+ *     void extendFinalizeFromProperties() override {
+ *         _coordDelay.set_delay(get_delay());
+ *         addComponent(&_coordDelay);
+ *     }
+ *     void extendConnectToModel(Model& model) override {
+ *         Super::extendConnectToModel(model);
+ *         const auto& coord = model.getCoordinateSet()[0];
+ *         // Set the input for the Delay component.
+ *         _coordDelay.getInput("input").connect(coord.getOutput("value"));
+ *     }
+ *
+ *     ScalarDelay _coordDelay;
+ * };
+ * @endcode
+ *
+ * This class is implemented via a SimTK::Measure_<T>::Delay.
+ *
+ * @ingroup operators
+ *
+ */
+template<typename T>
+class Delay : public Component {
+OpenSim_DECLARE_CONCRETE_OBJECT_T(Delay, T, Component);
+public:
+
+    /** @name Property declarations
+    These are the serializable properties associated with a Delay. **/
+    /**@{**/
+    OpenSim_DECLARE_PROPERTY(delay, double,
+        "The duration (nonnegative, seconds) by which the output is delayed.");
+    /**@}**/
+
+    /// Default constructor.
+    Delay();
+
+    /// Convenience constructor that sets the delay property.
+    explicit Delay(double delay);
+
+    /// Get the delayed value (the input's value at time t-delay).
+    T getValue(const SimTK::State& s) const;
+
+private:
+
+    void constructProperties() override;
+    void constructInputs() override;
+    void constructOutputs() override;
+
+    void extendFinalizeFromProperties() override;
+    void extendAddToSystem(SimTK::MultibodySystem& system) const override;
+
+    SimTK::MeasureIndex _delayMeasureIndex;
```

Comment:
I think it would be better here to make this a `SimTK::Measure_<T>`. That way you don't have to look it up at run time. `SimTK::Measure_<T>` is just a handle meaning it contains a pointer to the measure implementation object that lives in the multibody system.

Caution: copying will copy the measure handle (or measure index). Neither of these is going to be right; they should be cleared on copy like a ReferencePtr. Technically it shouldn't matter since a subsequent initSystem() call on the copy will set them correctly but it would be better if they were cleared on copy. Maybe we could make a ClearOnCopy<T> class that calls T.clear() on copy so that we can use it for a variety of reference objects?